src/main/java/org/truffleruby/core/binding/BindingNodes.java

```diff
@@ -179,6 +179,32 @@ public RubyNode coerceToString(RubyNode name) {
             return NameToJavaStringNodeGen.create(name);
         }
 
+        @Specialization(guards = {
+                "name == cachedName",
+                "!hiddenVariable(cachedName)",
+                "cachedFrameSlot != null",
+                "getFrameDescriptor(binding) == descriptor"
+        }, limit = "getCacheLimit()")
+        public Object localVariableDefinedCached(DynamicObject binding, String name,
+                @Cached("name") String cachedName,
+                @Cached("getFrameDescriptor(binding)") FrameDescriptor descriptor,
+                @Cached("findFrameSlotOrNull(name, getFrame(binding))") FrameSlotAndDepth cachedFrameSlot) {
+            return true;
+        }
+
+        @Specialization(guards = {
+                "name == cachedName",
+                "!hiddenVariable(cachedName)",
+                "cachedFrameSlot == null",
+                "getFrameDescriptor(binding) == descriptor"
+        }, limit = "getCacheLimit()")
+        public Object localVariableNotDefinedCached(DynamicObject binding, String name,
+                @Cached("name") String cachedName,
+                @Cached("getFrameDescriptor(binding)") FrameDescriptor descriptor,
+                @Cached("findFrameSlotOrNull(name, getFrame(binding))") FrameSlotAndDepth cachedFrameSlot) {
+            return false;
+        }
```

Comment:
This could be made a single @Specialization and then just return `cachedFrameSlot != null`